core/src/HIP/Kokkos_HIP_KernelLaunch.hpp

```diff
@@ -0,0 +1,232 @@
+                  ( hip_parallel_launch_local_memory
+                      < DriverType, MaxThreadsPerBlock, MinBlocksPerSM >
+                  , ( prefer_shmem ? hipFuncCachePreferShared :
+         hipFuncCachePreferL1 ) ) );
+            }
+            #endif
+
+            KOKKOS_ENSURE_CUDA_LOCK_ARRAYS_ON_DEVICE();*/
+
+      // Invoke the driver function on the device
+      printf("%i %i %i | %i %i %i | %i\n", grid.x, grid.y, grid.z, block.x,
+             block.y, block.z, shmem);
+      printf("Pre Launch Error: %s\n", hipGetErrorName(hipGetLastError()));
+
+      hipLaunchKernelGGL(
+          (hip_parallel_launch_local_memory<DriverType, MaxThreadsPerBlock,
+                                            MinBlocksPerSM>),
+          grid, block, shmem, hip_instance->m_stream, driver);
+
+      Kokkos::Experimental::HIP().fence();
+      printf("Post Launch Error: %s\n", hipGetErrorName(hipGetLastError()));
+#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
+      // CUDA_SAFE_CALL( hipGetLastError() );
+      Kokkos::HIP().fence();
+#endif
+    }
+  }
+  /*
+    static hipFuncAttributes get_hip_func_attributes() {
+      hipFuncAttributes attr;
+      hipFuncGetAttributes(&attr,hip_parallel_launch_local_memory
+              < DriverType, MaxThreadsPerBlock, MinBlocksPerSM >);
+      return attr;
+    }*/
+};
+
+template <class DriverType>
+struct HIPParallelLaunch<DriverType, Kokkos::LaunchBounds<0, 0>,
+                         HIPLaunchMechanism::LocalMemory> {
+  // static_assert(sizeof(DriverType)<HIPTraits::KernelArgumentLimit,"Kokkos
+  // Error: Requested HIPLaunchLocalMemory with a Functor larger than 4096
+  // bytes.");
+  inline HIPParallelLaunch(const DriverType &driver, const dim3 &grid,
+                           const dim3 &block, const int shmem,
+                           const HIPInternal *hip_instance,
+                           const bool prefer_shmem) {
+    if ((grid.x != 0) && ((block.x * block.y * block.z) != 0)) {
+      /**
+            if ( hip_instance->m_maxShmemPerBlock < shmem ) {
+              Kokkos::Impl::throw_runtime_exception(
+         std::string("HIPParallelLaunch FAILED: shared memory request is too
+         large") );
+            }
+
+            KOKKOS_ENSURE_CUDA_LOCK_ARRAYS_ON_DEVICE();
+      */
+      // Invoke the driver function on the device
+      hipLaunchKernelGGL(hip_parallel_launch_local_memory<DriverType>, grid,
+                         block, shmem, hip_instance->m_stream, driver);
+
+      Kokkos::Experimental::HIP().fence();
+#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
+      HIP_SAFE_CALL(hipGetLastError());
+      Kokkos::HIP().fence();
+#endif
+    }
+  }
+  /*
+    static hipFuncAttributes get_hip_func_attributes() {
+      hipFuncAttributes attr;
+      hipFuncGetAttributes(&attr,hip_parallel_launch_local_memory
+              < DriverType >);
+      return attr;
+    }*/
+};
+}  // namespace Impl
+}  // namespace Experimental
+}  // namespace Kokkos
+
+#endif
+
```

Comment:
So we don't have the constant cache or global launch mode yet?

Reply:
No because of a bug in `hipcc`. I reported the bug and told them it was urgent to fix it five months ago...

Reply:
Sounds good. Just wanted to know.